src/main/java/com/microfocus/application/automation/tools/octane/model/processors/scm/StarTeamSCMProcessor.java

```diff
@@ -0,0 +1,252 @@
+import com.hp.octane.integrations.dto.scm.SCMType;
+import com.microfocus.application.automation.tools.octane.configuration.SDKBasedLoggerProvider;
+import hudson.model.AbstractBuild;
+import hudson.model.Run;
+import hudson.model.User;
+import hudson.model.UserProperty;
+import hudson.scm.ChangeLogSet;
+import hudson.scm.EditType;
+import hudson.scm.SCM;
+import hudson.scm.ChangeLogSet.Entry;
+import hudson.tasks.Mailer;
+
+class StarTeamSCMProcessor implements SCMProcessor {
+  private static final Logger logger = SDKBasedLoggerProvider.getLogger(StarTeamSCMProcessor.class);
+  private static final DTOFactory dtoFactory = DTOFactory.getInstance();
+
+  @Override
+  public SCMData getSCMData(AbstractBuild build, SCM scm) {
+    List<ChangeLogSet<? extends ChangeLogSet.Entry>> changes = new ArrayList<>();
+    changes.add(build.getChangeSet());
+    return extractSCMData(build, scm, changes);
+  }
+
+  @Override
+  public SCMData getSCMData(WorkflowRun run, SCM scm) {
+    return extractSCMData(run, scm, run.getChangeSets());
+  }
+
+  @Override
+  public CommonOriginRevision getCommonOriginRevision(final Run run) {
+    return null;
+  }
+
+  private SCMData extractSCMData(Run run, SCM scm, List<ChangeLogSet<? extends ChangeLogSet.Entry>> changes) {
+    if (!(scm.getClass().getName().equals("hudson.plugins.starteam.StarTeamSCM"))) {    
+      throw new IllegalArgumentException("SCM type of StarTeamSCM was expected here, found '" + scm.getClass().getName() + "'");
+    }
+
+    SCMRepository repository = getRepository(scm);
+    List<SCMCommit> commits = extractCommits(changes);
+
+    return dtoFactory.newDTO(SCMData.class)
+        .setRepository(repository)
+        .setCommits(commits);
+  }
+
+  private SCMRepository getRepository(SCM starTeamSCM) {
+    SCMRepository result = null;
+    if (starTeamSCM != null) {
+      try {
+        String url = getSCMRepositoryURL(starTeamSCM);
+        String branch = getSCMRepositoryBranch(starTeamSCM);
+        result = dtoFactory.newDTO(SCMRepository.class).setType(SCMType.STARTEAM).setUrl(url).setBranch(branch);
+      } catch (IllegalAccessException | IllegalArgumentException | InvocationTargetException | NoSuchMethodException | SecurityException e) {
+        logger.warn(e.getClass().getSimpleName() + " unable to extract data from starTeamSCM, no SCM repository info will be available.", e);
+      }
+    } else {
+      logger.warn("starTeamSCM is null, no SCM repository info will be available.");
+    }
+    return result;
+  }
+  
+  private List<SCMCommit> extractCommits(List<ChangeLogSet<? extends ChangeLogSet.Entry>> changes) {
+    LinkedList<SCMCommit> tmpCommits = new LinkedList<>();
+  
+    for (ChangeLogSet<? extends ChangeLogSet.Entry> set : changes) {
+      for (ChangeLogSet.Entry change : set) {
+        if (change.getClass().getName().equals("hudson.plugins.starteam.changelog.StarTeamChangeLogEntry")) {
+          List<SCMChange> tmpChanges = new ArrayList<>();
```

Comment:
Can you try to rewrite this? For loop inside for loop inside another for loop..

Reply:
I prefer to leave it as is because it's not so complex as it sound.